chain/actions/src/payload.rs

```diff
@@ -213,51 +224,155 @@ impl PayloadGenerator<TransactionRequest> for BasicPayloadGenerator {
+        &self,
+        destinations: S,
+    ) -> Result<TransactionRequest> {
+        let destinations = destinations.into_iter().collect::<Vec<_>>();
+        match destinations.len() {
+            0 => Err(InvalidArguments("No outgoing channel to initiate closure".into())),
+            1 => {
+                let destination = destinations.into_iter().next().unwrap();
+                if destination.eq(&self.me) {
+                    return Err(InvalidArguments(
+                        "Cannot initiate closure of outgoing channel from self".into(),
+                    ));
                 }
-                .abi_encode(),
-            )
-            .with_to(self.contract_addrs.channels.into());
-        Ok(tx)
-    }
 
-    fn finalize_outgoing_channel_closure(&self, destination: Address) -> Result<TransactionRequest> {
-        if destination.eq(&self.me) {
-            return Err(InvalidArguments(
-                "Cannot initiate closure of incoming channel to self".into(),
-            ));
+                let tx = TransactionRequest::default()
+                    .with_input(
+                        initiateOutgoingChannelClosureCall {
+                            destination: destination.into(),
+                        }
+                        .abi_encode(),
+                    )
+                    .with_to(self.contract_addrs.channels.into());
+                Ok(tx)
+            }
+            _ => {
+                // create multisend data payload
+                let multisend_txns = destinations
+                    .into_iter()
+                    .filter(|destination| destination != &self.me) // filter the outgoing channel to self
+                    .map(|destination| {
+                        let encoded_data = initiateOutgoingChannelClosureCall {
+                            destination: destination.into(),
+                        }
+                        .abi_encode();
+                        MultisendCallOnlyTransaction {
+                            encoded_data: encoded_data.into(),
+                            to: self.contract_addrs.channels.into(),
+                            value: U256::ZERO,
+                        }
+                    })
+                    .collect();
+
+                let tx_payload = MultisendCallOnlyTransaction::build_multisend_tx(multisend_txns);
+                debug!(?tx_payload, "Multisend payload for initiating outgoing channel closure");
+
+                let tx = TransactionRequest::default()
+                    .with_input(tx_payload)
+                    .with_to(MULTISEND_CALL_ONLY);
+                Ok(tx)
+            }
         }
+    }
 
-        let tx = TransactionRequest::default()
-            .with_input(
-                finalizeOutgoingChannelClosureCall {
-                    destination: destination.into(),
+    fn finalize_outgoing_channel_closure<S: IntoIterator<Item = Address>>(
```

Comment:
This is a lot of copy-pasted logic and code, please, extract the common functionality of the code into a single helper.

Reply:
You'd have something like: perform_action<TypeOfAction>(data: IntoIter)